File: workbook/ch07/data/buffer/circular.py
class CircularBuffer:
    def __init__(self, size):
        self.size = size
        self.buffer = [None] * size
        self.head = 0  # Next insert
        self.tail = 0  # Next remove
        self.count = 0

    def is_full(self):
        return self.count == self.size

    def is_empty(self):
        return self.count == 0

    def enqueue(self, value):
        if self.is_full():
            return False
        self.buffer[self.head] = value
        self.head = (self.head + 1) % self.size
        self.count += 1
        return True

    def dequeue(self):
        if self.is_empty():
            return None
        value = self.buffer[self.tail]
        self.tail = (self.tail + 1) % self.size
        self.count -= 1
        return value

    def __str__(self):
        elements = [
            str(self.buffer[(self.tail + i) % self.size])
            for i in range(self.count)
        ]
        return "Buffer: [" + " ".join(elements) + "]"

if __name__ == "__main__":
    cb = CircularBuffer(5)

    cb.enqueue(10)
    cb.enqueue(20)
    cb.enqueue(30)
    cb.enqueue(40)
    cb.enqueue(50)

    print(cb)

    print("Dequeued:", cb.dequeue())  # 10
    cb.enqueue(60)

    print(cb)

    while not cb.is_empty():
        print("Dequeued:", cb.dequeue())

    print("Final buffer state:", cb)

File: workbook/ch07/data/buffer/circular/circular.c
#include <stdio.h>
#include <stdbool.h>

#define BUFFER_SIZE 5

typedef struct {
    int data[BUFFER_SIZE];
    int head;  // points to the next insertion index
    int tail;  // points to the next removal index
    int count; // number of elements currently in buffer
} CircularBuffer;

void init(CircularBuffer *cb) {
    cb->head = 0;
    cb->tail = 0;
    cb->count = 0;
}

bool is_full(CircularBuffer *cb) {
    return cb->count == BUFFER_SIZE;
}

bool is_empty(CircularBuffer *cb) {
    return cb->count == 0;
}

bool enqueue(CircularBuffer *cb, int value) {
    if (is_full(cb)) return false;

    cb->data[cb->head] = value;
    cb->head = (cb->head + 1) % BUFFER_SIZE;
    cb->count++;
    return true;
}

bool dequeue(CircularBuffer *cb, int *value) {
    if (is_empty(cb)) return false;

    *value = cb->data[cb->tail];
    cb->tail = (cb->tail + 1) % BUFFER_SIZE;
    cb->count--;
    return true;
}

void print_buffer(CircularBuffer *cb) {
    printf("Buffer: ");
    for (int i = 0; i < cb->count; i++) {
        int index = (cb->tail + i) % BUFFER_SIZE;
        printf("%d ", cb->data[index]);
    }
    printf("\n");
}




int main() {
    CircularBuffer cb;
    init(&cb);

    enqueue(&cb, 10);
    enqueue(&cb, 20);
    enqueue(&cb, 30);
    enqueue(&cb, 40);
    enqueue(&cb, 50);  // fills buffer

    print_buffer(&cb);

    int val;
    dequeue(&cb, &val); // removes 10
    printf("Dequeued: %d\n", val);

    enqueue(&cb, 60); // wraps around
    print_buffer(&cb);

    while (dequeue(&cb, &val)) {
        printf("Dequeued: %d\n", val);
    }

    return 0;
}